op-node: Span Batch Type, Encoding, and Decoding

ImTei opened the pull request:
## Contexts

This PR contains the type, encoding, and decoding code for [Span Batch](https://github.com/ethereum-optimism/optimism/blob/develop/specs/span-batches.md) as defined specs. 

Please refer to the Encoding Optimization section in Design Docs for the rationale behind the current encoding specs.

This PR has a dependency on the[ **Hardfork Configuration** PR](https://github.com/ethereum-optimism/optimism/pull/7287).

## Changes

- Added the following types:
    - Batch interface
    - RawSpanBatch
    - SpanBatchElement
    - SpanBatch
    - SpanBatchBuilder
- Changed the following types:
    - Renamed BatchV1 to SingularBatch.
    - BatchData
- ^ Please refer to the `Type` section of Implementation Design Docs for details.
- Added unit tests.
    - ^ Please refer to the `Data Structure` category in the Test List Sheet for each test’s details.

Commit:
Implement Span batch encoding and decoding

Define SpanBatch and related types
Rename BatchV1 to SingularBatch
Add unit test cases

protolambda commented on op-node/rollup/derive/batch.go:
```diff
@@ -7,48 +7,44 @@ import (
-
-func (b *BatchV1) Epoch() eth.BlockID {
-	return eth.BlockID{Hash: b.EpochHash, Number: uint64(b.EpochNum)}
+	BatchType int
+	SingularBatch
+	RawSpanBatch
```
Adding the type explicitly, and embedding each kind, is probably the less intrusive change we can make to the existing approach with typed BatchData. We can maybe refactor it to be more similar to `types.Transaction` in upstream geth, where a struct is used to represent the typed decoding/encoding and any caching, and wraps around a single interface, rather than embedding each type of possible inner batch. We can keep this for now, to keep the PR stack easier to land into develop.

pcw109550 replied:
Will do this in a separate PR. Strongly agree that following what `types.Transaction` did is better because current composition pattern will become more complicated when we need another sub field of composition. We need to add additional field for new span batch type when needed if we do not use a single interface.

Let me make the abstraction idea concrete. type `BatchData` will be refactored as below:
```go
type BatchData struct {
	inner BatchInnerData
}
```
where `BatchInnerData` is an interface:
```go
type BatchInnerData interface {
	batchType() int
        encode(w io.Writer) error
        decode(r *bytes.Reader) error
}
```

Each `SingularBatch` and `RawSpanBatch` will implement `BatchInnerData` interface. This will encapsulate all the batch type related conditionals to each implementations: for example below method will be refactored from:
```go
func (b *BatchData) encodeTyped(buf *bytes.Buffer) error {
	switch b.BatchType {
	case SingularBatchType:
		buf.WriteByte(SingularBatchType)
		return rlp.Encode(buf, &b.SingularBatch)
	case SpanBatchType:
		buf.WriteByte(byte(b.BatchType))
		return b.RawSpanBatch.encode(buf)
	default:
		return fmt.Errorf("unrecognized batch type: %d", b.BatchType)
	}
}
```
to
```go
func (b *BatchData) encodeTyped(buf *bytes.Buffer) error {
        if err := buf.WriteByte(b.inner.batchType()); err != nil {
            return err
        }
        return b.inner.encode(buf)
}
```

pcw109550 replied:
I just found out that `buf.WriteByte(SingularBatchType)` is not error handled. Will also fix this.

protolambda commented on op-node/rollup/derive/batch_tob_test.go:
```diff
@@ -10,6 +10,7 @@ import (
 
 // FuzzBatchRoundTrip executes a fuzz test similar to TestBatchRoundTrip, which tests that arbitrary BatchData will be
 // encoded and decoded without loss of its original values.
+// Does not test span batch because fuzzer is not aware structure of span batch.
```
```suggestion
// Does not test the span batch type because the fuzzer is not aware of the structure of a span batch.
```

protolambda commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,670 @@
+type spanBatchPrefix struct {
+	relTimestamp  uint64 // Relative timestamp of the first block
+	l1OriginNum   uint64 // L1 origin number
+	parentCheck   []byte // First 20 bytes of the first block's parent hash
+	l1OriginCheck []byte // First 20 bytes of the last block's L1 origin hash
```
Stylistically we can make these fields `[20]byte`, but also fine to leave it the same for now, functionally the same.

protolambda commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,670 @@
+
+// LogContext creates a new log context that contains information of the batch
+func (b *SpanBatch) LogContext(log log.Logger) log.Logger {
+	return log.New(
```
Although empty span-batches are invalid, we might still log a batch while we're building it locally using an empty span batch as start + append calls. Adding a check here in case `len(b.batches) == 0` to create a different log context safely, and not cause a panic, would be nice.

protolambda commented on op-node/rollup/derive/span_batch_test.go:
```diff
@@ -0,0 +1,514 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/stretchr/testify/assert"
```
In some cases, when asserting if an error is not nil, using the `require` package (same API) is better, to stop the test at that unexpected error rather than continuing. E.g. `require.NoError(err)`. It can be fixed in a follow-up PR though, it's fine to keep as-is for now.

protolambda commented on op-node/rollup/derive/span_batch_tx.go:
```diff
@@ -0,0 +1,253 @@
+	}
+}
+
+// DecodeRLP implements rlp.Decoder
+func (tx *spanBatchTx) DecodeRLP(s *rlp.Stream) error {
```
In `func (btx *spanBatchTxs) fullTxs(chainID *big.Int)` you use the `func (tx *spanBatchTx) UnmarshalBinary(b []byte) error {` version. Is the `DecodeRLP` method really required? And `newSpanBatchTxs` uses `MarshalBinary` too.

ImTei replied:
`DecodeRLP` was implemented just for convention with other types. It's not required. Let me remove it in a follow-up PR!

protolambda commented on op-node/rollup/derive/span_batch_txs.go:
```diff
@@ -0,0 +1,475 @@
+		var tx types.Transaction
+		if err := tx.UnmarshalBinary(txs[idx]); err != nil {
+			return nil, errors.New("failed to decode tx")
+		}
```
```suggestion
		}
		if tx.Protected() && tx.ChainId().Cmp(chainID) != 0 {
			return nil, fmt.Errorf("protected tx has chain ID %d, but expected chain ID %d", tx.ChainId(), chainID)
		}
```

The `chainID` argument is unused. We can use it for a sanity check like this. (but only if the tx is protected, i.e. is specific to a single chain).

protolambda commented:
Adding a do-not-merge label to give others the chance to review before auto-merge by mergify kicks in.

protolambda reviewed (approved):
Looks good! There are some stylistic but breaking things I think we can do in a follow-up PR, after the PR stack is wrapped up.

trianglesphere commented on op-node/rollup/derive/batch.go:
```diff
@@ -69,9 +65,18 @@ func (b *BatchData) MarshalBinary() ([]byte, error) {
+		buf.WriteByte(SingularBatchType)
+		return rlp.Encode(buf, &b.SingularBatch)
+	case SpanBatchType:
+		buf.WriteByte(byte(b.BatchType))
```
```suggestion
		buf.WriteByte(SpanBatchType)
```

trianglesphere commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,670 @@
+// SpanBatchType := 1
+// spanBatch := SpanBatchType ++ prefix ++ payload
+// prefix := rel_timestamp ++ l1_origin_num ++ parent_check ++ l1_origin_check
+// payload := payload = block_count ++ origin_bits ++ block_tx_counts ++ txs
```
```suggestion
// payload := block_count ++ origin_bits ++ block_tx_counts ++ txs
```

trianglesphere commented on op-node/rollup/derive/span_batch.go:
```diff
@@ -0,0 +1,670 @@
+
+// decodeParentCheck parses data into bp.parentCheck
+func (bp *spanBatchPrefix) decodeParentCheck(r *bytes.Reader) error {
+	bp.parentCheck = make([]byte, 20)
```
A lot of the decode functions are duplicating basic building blocks like decode20Bytes or decodeUVarInt

pcw109550 replied:
Code duplication should be avoided, but these structures are intentionally duplicated and not abstracted. For example, lets say we want to abstract out `binary.ReadUvarint(r)` as a function `decodeUvarint`:
```go
func decodeUvarint(r *bytes.Reader) (uint64, error) {
    return binary.ReadUvarint(r)
}
```
Above implementation is an unncessary abstraction. `decode20Bytes` would be better:
```go
func decode20Bytes(r *bytes.Reader) ([]byte, error) {
	result := make([]byte, 20)
	_, err := io.ReadFull(r, result)
	if err != nil {
		return nil, err
	}
	return result, err
}
```
and implement `decodeParentCheck` like:
```go
func (bp *spanBatchPrefix) decodeParentCheck(r *bytes.Reader) error {
	result, err := decode20Bytes(r)
	if err != nil {
		return fmt.Errorf("failed to read parent check: %w", err)
	}
	bp.parentCheck = result
	return nil
}
```
However I can observe the pattern: call `make` to allocate and call `io.ReadFull` throughout the entire codebase. These patterns are not abstracted, and I think leaving it as is(not encapsulating as function) is better for readability.

pcw109550 replied:
Also, we can observer structure of method `decodeParentCheck` and `decodeL1OriginCheck` is identical:
```go
// decodeParentCheck parses data into bp.parentCheck
func (bp *spanBatchPrefix) decodeParentCheck(r *bytes.Reader) error {
	bp.parentCheck = make([]byte, 20)
	_, err := io.ReadFull(r, bp.parentCheck)
	if err != nil {
		return fmt.Errorf("failed to read parent check: %w", err)
	}
	return nil
}

// decodeL1OriginCheck parses data into bp.decodeL1OriginCheck
func (bp *spanBatchPrefix) decodeL1OriginCheck(r *bytes.Reader) error {
	bp.l1OriginCheck = make([]byte, 20)
	_, err := io.ReadFull(r, bp.l1OriginCheck)
	if err != nil {
		return fmt.Errorf("failed to read l1 origin check: %w", err)
	}
	return nil
}

pcw109550 replied:
Also, we can observe structure of method `decodeParentCheck` and `decodeL1OriginCheck` is identical:
```go
// decodeParentCheck parses data into bp.parentCheck
func (bp *spanBatchPrefix) decodeParentCheck(r *bytes.Reader) error {
	bp.parentCheck = make([]byte, 20)
	_, err := io.ReadFull(r, bp.parentCheck)
	if err != nil {
		return fmt.Errorf("failed to read parent check: %w", err)
	}
	return nil
}

// decodeL1OriginCheck parses data into bp.decodeL1OriginCheck
func (bp *spanBatchPrefix) decodeL1OriginCheck(r *bytes.Reader) error {
	bp.l1OriginCheck = make([]byte, 20)
	_, err := io.ReadFull(r, bp.l1OriginCheck)
	if err != nil {
		return fmt.Errorf("failed to read l1 origin check: %w", err)
	}
	return nil
}
```
and I may implement `decode20Bytes` function and call it twice instead of calling different two methods. I thought that even though there is a duplication, breaking the decoding logic into chunks and explicitly naming them(make them into function) would be much better for readability.

ImTei committed:
Apply suggestions from code reviews

Fix comments
Check if the span batch is empty in LogContext()
Check chain ID of the protected tx in newSpanBatchTxs()
Write SpanBatchType explicitly when encoding span batch

ImTei committed:
Update testutils path

ImTei commented:
applied simple [changes](https://github.com/ethereum-optimism/optimism/pull/7288/commits/23c658d60e03e633a9097e5939d33798f2837170) from code reviews and rebased to the latest develop branch.

Other suggestions will be applied in the follow-up PR!

protolambda commented on op-node/rollup/derive/singular_batch.go:
```diff
@@ -0,0 +1,53 @@
+// Batch format
+//
+// SingularBatchType := 0
+// singularBatch := SingularBatchType ++ RLP([epoch, timestamp, transaction_list]
```
```suggestion
// singularBatch := SingularBatchType ++ RLP([parent_hash, epoch_number, epoch_hash, timestamp, transaction_list])
```

protolambda committed:
op-node: correct batch-encoding doc-comment

protolambda commented:
Follow-ups to implement in later PR:

- `spanBatchTxs` has unused `EncodeRLP` and `DecodeRLP`. The marshal/unmarshal-binary functions are enough.
- Some encoding/decoding tests in `span_batch_tx_test.go` randomize the test data and assert 
  if the tx types all occurred at the end. Instead, we can make sub-tests with `t.Run()` for each tx-type,
  and randomize the same amount of txs per type.
- The `BatchData` wrapper around the typed batch shouldn't have an explicit byte field for the type:
  it can already be inferred from the contained batch data. See `types.Transaction` in geth as example.
- `decodeTyped` can be simplified, and the switch-statement should have be more consistent (not mix constants and `data[0]`)
- `SingularBatchType = iota` is neat, but being explict about protocol constants is better. `iota` is only really meant for internal enumerations that you can flexibly add/remove to.
- Review usage of `require` vs `assert` in batch tests
- `parentCheck   []byte` and `l1OriginCheck []byte` should be `[20]byte` types

And opened #7615 to fix a blocking issue: we must not allow span-batches to be decoded successfully before the HF, or we diverge from the actual mainnet specs.
